tuf/client/updater.py

```diff
@@ -650,89 +650,73 @@ def refresh(self, unsafely_update_root_if_necessary=True):
-
-      else:
-        raise
-
-    except tuf.ExpiredMetadataError:
-      if unsafely_update_root_if_necessary:
-        logger.info('No changes were detected from the mirrors for a given role'
-          ', and that metadata that is available on disk has been found to be'
-          ' expired. Trying to update root in case of foul play.')
-        self._update_metadata('root', DEFAULT_ROOT_UPPERLENGTH)
-        self.refresh(unsafely_update_root_if_necessary=False)
-
-      # The caller explicitly requested not to unsafely fetch an expired Root.
-      else:
-        logger.info('No changes were detected from the mirrors for a given role'
-          ', and that metadata that is available on disk has been found to be '
-          'expired. Your metadata is out of date.')
-        raise
-
-
+    self._update_metadata('timestamp', DEFAULT_TIMESTAMP_UPPERLENGTH)
+    self._update_metadata_if_changed('snapshot',
+                                     referenced_metadata='timestamp')
+    self._update_metadata_if_changed('targets')
+    
 
+  def _update_root_metadata(self, current_root_metadata, compression_algorithm=None):
+    """
+      <Purpose>
+        The root file must be signed by the current root threshold and keys as well
+        as the previous root threshold and keys. The update process for root files 
+        means that each intermediate root file must be downloaded, to build a chain 
+        of trusted root keys from keys already trusted by the client:
+        
+          1.root -> 2.root -> 3.root
+         
+        3.root must be signed by the threshold and keys of 2.root, and 2.root must
+        be signed by the threshold and keys of 1.root.
+    
+      <Arguments>
+        current_root_metadata:
+          The currently held version of metadata.
+           
+        compresison_algorithm:
+          The compression algorithm used to compress remote files.
+    
+      <Side Effects>
+        Updates the root metadata files with the latest information.
+    
+      <Returns>
+        None.
+    """
+    next_version = current_root_metadata['version'] + 1
+    new_root_metadata_file = self._get_metadata_file('root', 'root.json', tuf.conf.DEFAULT_ROOT_REQUIRED_LENGTH, None, 
+                                                     compression_algorithm=compression_algorithm)
+    new_root_metadata = \
+      tuf.util.load_json_string(new_root_metadata_file.read().decode('utf-8'))
+    new_version = new_root_metadata['signed']['version'] + 1
+    
+    for version in range(next_version, new_version):
+      # Temporarily set consistent snapshot. Will be updated to whatever is set in the latests root.json after 
+      # running through the intermediates with `_update_metadata`
+      self.consistent_snapshot = True
+      self._update_metadata('root', tuf.conf.DEFAULT_ROOT_REQUIRED_LENGTH, version=version, 
+                            compression_algorithm=compression_algorithm)
```

Comment:
According to the new update process, top-level metadata is now downloaded in `root -> timestamp -> snapshot -> targets -> (delegations)` order.  When updating root, the client updates root from the current version up to the latest (including intermediate versions).  Supposing the current version = 1 and the latest version = 5, the client fetches 1.root.json -> 2.root.json -> ... -> 5.root.json.  Currently, root.json is downloaded if any of the top-level metadata cannot be successfully updated, and immediately (if needed) after updating snapshot.json.  The snapshot.json role lists root.json, and timestamp.json provides timeliness for all the other top-level metadata.

As I understand the modified update process, lines 707 - 719 attempt to fetch the latest root to determine the `new_version`.  One problem I see with this approach is that the fetched root file is not verified.  Is this really the latest version?  Since root is not updated after snapshot.json nor its latest version referenced from snapshot.json, I assume the "latest" version of root fetched remotely is treated as up-to-date, until it expires?